### PR TITLE
Add powerEfficientPixelFormat

### LINKED ISSUE #13: Add constraint to avoid inefficient (compressed) pixel formats
Capturing in compressed pixel formats such as MJPEG adds CPU overhead because the browser has to convert (decompress) every frame before delivery to the MediaStreamTrack and beyond.

An application that cares about both quality and performance might ask with non-required constraints for Full HD. If the user has a USB 3.0 camera, Full HD might be delivered without any compression overhead. Great! But if the user has a USB 2.0 camera, due to bus limitations, Full HD would (on cameras available for testing) be captured in MJPEG, adding this overhead. The application pays a performance debt, even though it might have been just as happy if it got HD frames at a lower cost.

TL;DR: Should we add a `{video:{avoidCapturingExpensivePixelFormats:true}}` constraint? I'm not married to the name :)

## Motivation in Numbers

Frames are captured in one format, typically NV12 (420v), YUY2 (yuvs) or MJPEG (dmb1) and then converted. Chromium traditionally converts to I420 (y420) as this format is widely supported by encoders, though it is possible to have other destination pixel formats (e.g. NV12 is supported by some encoders which could allow for a zero-conversion pipeline in WebRTC).

While YUY2 to I420 is fairly cheap, MJPEG to I420 isn't as cheap.

I set up thin "capture and convert" demo ([code](https://chromium-review.googlesource.com/c/chromium/src/+/2450194)) and measured the CPU usage (utilization percentage normalized by CPU frequency using Intel Power Gadget and a [script](https://github.com/henbos/power-gadget) to obtain a sense of the "absolute" amount of work performed).

Here is the result of capturing in various formats*, converting to I420 at 30 fps and measuring the CPU and power consumption.

_* Caveat: NV12 and YUY2 are captured with the built-in MacBook Pro camera and MJPEG is captured using an external Logitech Webcam C930e. The external webcam could contribute to some of the added CPU usage and power consumption, so it would be good to compare MJPEG on webcam with YUY2 on the same webcam, but the majority of the work is in the pixel conversions._

Capture Format | Resolution | Normalized CPU Usage [M cycles/s] | Power Consumption [Watt]
-- | -- | -- | --
NV12 (420v) | 640x480 (VGA) | 26.51 | 3.10
... | 1280x720 (HD) | 28.94 | 3.23
YUY2 (yuvs) | 640x480 (VGA) | 20.57 | 2.98
... | 1280x720 (HD) | 30.97 | 3.31
MJPEG (dmb1) | 640x480 (VGA) | 52.85 | 4.85
... | 1280x720 (HD) | 67.99 | 5.28
... | 1920x1080 (Full HD) | 102.41 | 6.27

Note: I am not measuring the entire browser, I am only measuring a demo that does capturing and conversion.

In this example...
- Capturing in YUY2 at HD instead of MJPEG at Full HD reduces the CPU usage from 102.41 to 30.97 M cycles/s (-70%) and the power consumption from 6.27 to 3.31 W (-47%). We save ~3 W.
- Capturing in YUY2 at HD instead of MJPEG at HD reduces the CPU usage from 67.99 to 30.97 M cycles/s (-54%) and the power consumption from 5.28 to 3.31 W (-37%). We save ~2 W.

## Proposal

Add a new video constraint, e.g. BooleanConstraint avoidCapturingExpensivePixelFormats, that if true allows the browser to skip pixel formats of a device that are deemed inefficient (e.g. MJPEG) if that same device supports capturing in other pixel formats.

On Logitech Webcam C930e, where Full HD is only available as MJPEG but 1024x576 and below is available as YUY2, getUserMedia would pick a lower resolution but avoid MJPEG.

P.S. This could result in a tradeoff between frame rate and resolution, more discussion needed.

### PR DESCRIPTION
**Pull request by henbos**
Fixes #13.

Add a constraint to allow the application to influence the performance/quality tradeoff. For example we can require the the pixel format is efficient. In practise this could be used to avoid MJPEG decoding load of high resolution USB 2.0 cameras while still allowing high resolution uncompressed NV12 on USB 3.0. The application does not have to care about specific pixel formats, but it can constrain on "efficiency".

***
<a href="https://pr-preview.s3.amazonaws.com/henbos/mediacapture-extensions/pull/59.html" title="Last updated on May 20, 2022, 2:07 PM UTC (cea2c4a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-extensions/59/799b509...henbos:cea2c4a.html" title="Last updated on May 20, 2022, 2:07 PM UTC (cea2c4a)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -208,6 +208,31 @@ <h3>GetUserMediaSemantics enum</h3>
...
+      <div>
+        <pre class="idl"
+>partial dictionary MediaTrackConstraintSet {
+  ConstrainBoolean efficientPixelFormat;
+};</pre>
+        <section>
+          <h2>Dictionary {{MediaTrackConstraintSet}} Members</h2>
+          <dl data-link-for="MediaTrackConstraintSet" data-dfn-for=
+          "MediaTrackConstraintSet" class="dictionary-members">
+            <dt><dfn><code>efficientPixelFormat</code></dfn> of type <span class=
```
Comment (handellm) [inline index.html:223]: I would suggest s/efficientPixelFormat/powerEfficient/ to be more like https://www.w3.org/TR/media-capabilities/#media-capabilities-info.

File: index.html [hunk 2]
```diff
@@ -208,6 +208,31 @@ <h3>GetUserMediaSemantics enum</h3>
...
+            "idlMemberType">{{ConstrainBoolean}}</span></dt>
+            <dd>
+              <p>It is common for cameras to support different pixel formats for
```
Comment (handellm) [inline index.html:226]: Maybe talk less of pixel formats? Something like
"Some capture modes are more expensive than others when it comes to power consumption. The user agent SHOULD label modes that incur significant power penalty as not power efficient. What format is or is not considered power efficient is up to the user agent, but modes that cause software MJPEG decode is an example of expensive modes."

**Comment by jan-ivar** [comment]
> For example we can require the the pixel format is efficient.

When you say "require", I'm not a fan of sites restricting what cameras can be used, as this limits user choice.

But luckily, this PR doesn't do that, since you've not modified the [allowed required constraints for device selection](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-allowed-required-constraints-for-device-selection).

> In practise this could be used to avoid MJPEG decoding load of high resolution USB 2.0 cameras while still allowing high resolution uncompressed NV12 on USB 3.0. T

So:
* `{height: 1080, powerEfficient: true}` might yield <1080p if it avoids MJPEG over USB 2.0?
* `{height: 1080, powerEfficient: false}` wouldn't?
* `{height: 1080}` would do what exactly?

Avoiding power-inefficient modes of a user-chosen camera seems desirable. But why would a user agent need app input to do this? If it has this ability and knowledge, shouldn't it apply it by default? And if it's the default, isn't it better for the app to constrain on what it _wants_, instead of what it doesn't want? E.g. `{height: {min: 1080}}` for quality over performance? That way the browser can do the power-efficient thing when it can, and only resort to other modes when it has to to meet required app constraints.

The answer to this question may be that it's not web compatible at this point. Just wanted to raise it.

**Comment by youennf** [comment]
`powerEfficient` seems to be something difficult to define so I am not sure I like it.
For instance `{height: 1080, powerEfficient: true}` could lead to choosing a frame rate of 1 frame per second?

I would also echo Jan-Ivar question about UA trying to choose the most optimised pixel format it can by default.
The question is: Are some web apps interested in selecting non optimised format (MJPEG say might be handy in some cases)?

If so, an API that may be useful would be something like:
- Open the camera, apply some settings.
- Expose some additional settings that can be set, including available native pixel formats.
- Let the web application decide which native pixel format it wants (the default being the UA preferred one).

**Comment by henbos** [comment]
Reply to @jan-ivar 

> Avoiding power-inefficient modes of a user-chosen camera seems desirable. But why would a user agent need app input to do this? If it has this ability and knowledge, shouldn't it apply it by default? And if it's the default, isn't it better for the app to constrain on what it wants, instead of what it doesn't want?

The rationale for adding an additional constraint was this: "How can the user agent know what to prefer (quality or performance) if the app does not say? And how can the app say what it wants, if it cannot say what to prefer (quality or performance)?"

But it is true, the user agent could default to preferring performance such that if you ask for ideal:1080p you'd get 480p or 720p based on efficiency. In practise 720p would also be an inefficient format on USB 2.0, but limiting the app to 480p seems a bit excessive for someone asking for 1080p!

> The answer to this question may be that it's not web compatible at this point. Just wanted to raise it.

It's hard to imagine not causing web compatibility issues without a new constraint. And the fitness distance algorithm would become less relevant if user agents started to add its own fitness distance penalty based on performance into the mix. We might be able to solve it without a constraint like you say, but I'm not sure it's the path we want to go.

> So:
>
> - {height: 1080, powerEfficient: true} might yield <1080p if it avoids MJPEG over USB 2.0?

Wouldn't {height:1080} ensure exact 1080 or fail? But ideal height and exact powerEfficient might yield <1080p, yes.
However to avoid 480p you might want to use advanced constraints to ensure powerEfficient only plays a role in Full HD. (That or multiple GUM calls?)

**Comment by henbos** [comment]
Reply to @youennf 

> For instance {height: 1080, powerEfficient: true} could lead to choosing a frame rate of 1 frame per second?

Yes. You would have to be careful with powerEfficient since high resolution efficient formats may have worse fps.
To avoid frame rate degration, it would be wise to specify a min frameRate as well.

> If so, an API that may be useful would be something like:
>
> - Open the camera, apply some settings.
> - Expose some additional settings that can be set, including available native pixel formats.
> - Let the web application decide which native pixel format it wants (the default being the UA preferred one).

If we gave the app full control of camera capture formats then we might want to go down this route. But I think that is a separate issue.

But for now, I don't think the app would or should care about specific pixel formats. It should only care if it's powerEfficient or not. Even the big 'baddie' MJPEG could be powerEfficient if there is a hardware decoder available, and I don't think we need to expose this level of detail to the app for what I am trying to solve here.

**Comment by henbos** [comment]
> powerEfficient seems to be something difficult to define so I am not sure I like it.

Efficiency is hard to define, but that doesn't mean it's not useful (the [other powerEfficient](https://www.w3.org/TR/media-capabilities/#media-capabilities-info) being a prime example of this). I still think that a separate constraint is more easy to define than fitness distance penalties.

**Review by henbos — COMMENTED** [review]
Addressed comments

**Comment by henbos** [comment]
Show of hands?

Proposal A:
- powerEfficient is added, allowing the app to say what it prefers and feature detect.

Proposal B:
- No new constraint, it's entirely up to the browser. Either we say nothing at all or we add a note suggesting the browser "should" prefer performance.
- If an app wants to ensure high resolution, it can do so with advanced constraints or multiple GUM calls.

**Comment by youennf** [comment]
> With fitness distance, we worked hard to walk back the constraints language from something extremely complicated (min/max/exact/advanced) to something more declarative.

Adding a powerEfficient constraint is extremely complex and does not scale well.
First the more you add constraints, the less predictable the fitness distance makes sense (adding carrots and leeks...).
Second, powerEfficient is a very loose term (at least in MC, we have a common understanding about HW support).
I do not expect A to be a viable interoperable solution.

There are alternatives we could consider by splitting powerEfficient in two:

1. Expose available natively supported pixel formats.
The good news is that we already have a definition of pixel formats at https://www.w3.org/TR/webcodecs/#pixel-format.
It might be ok to expose it as a constraint (sigh... mainly for applyConstraints use though).

2. Allow the web application to select a set of constraint values that perfectly matches a natively supported camera configuration.
To mimic what OS APIs are doing (a good thing in general), we can expose presets like discussed in https://github.com/w3c/mediacapture-extensions/issues/12 (which does mention exposing pixel format).
Presets would include available camera native pixel formats at native resolution/frame rates.
'powerEfficient' is then defined by web applications according what the camera can offer and what the web application needs, not by user agents.

**Comment by henbos** [comment]
That would be Proposal C: Expose camera presets to web pages.

The app would not care about VideoPixelFormat per se, only if it is or isn't avoiding software decoding in the conversion between native capture and whatever the rest of the UA pipeline is using. So I would suspect that even in Proposal C, you'd end up with something like powerEfficient in the presets anyway. In which case "presets versus constraints" becomes a tangental discussion to whether or not powerEfficient is a good idea.

> Second, powerEfficient is a very loose term (at least in MC, we have a common understanding about HW support).

Would you still oppose this if the definition was as well defined as MC's powerEfficient?
I'd be happy with `powerEfficient: This capture format avoids software decoding of MJPEG`.

**Comment by youennf** [comment]
> Would you still oppose this if the definition was as well defined as MC's powerEfficient?

I do not think this is possible.
powerEfficient for MC is restricted to the encoder or decoder. That would mirror here to scoping the definition to cameras.
Cameras that are outputting MJPEG can be considered power efficient if they are doing this natively, in HW say.

Your definition of powerEfficient encompasses the whole video pipeline, not just the camera.
If an application does media recording in JS using MJPEG directly, this qualifies as power efficient.

> I'd be happy with `powerEfficient: This capture format avoids software decoding of MJPEG`.

If what you are after is avoiding MJPEG video frames, a dedicated API (do not use MJPEG) can be built for that.
This seems like an edge case that web developers will not be aware of and where the UA default is of prime importance.
I would not use the term powerEfficient for that API.

Exposing pixel format selection instead seems to fix the issue and would be useful outside of this edge case.
Now that we are exposing camera pixel format with media capture transform to JS, the actual pixel format being used by cameras might become an interop issue.

> So I would suspect that even in Proposal C, you'd end up with something like powerEfficient in the presets anyway

Not really. If a preset is used, it will be powerEfficient with regards to the camera.
It is then up to the web application, that is designing the video pipeline, to see whether the camera preset output matches the rest of the pipeline.
If the pipeline is RTCPeerConnection, YUV is probably the best choice.
If the pipeline is WebCodecs, MC will tell what pixel format seems best.
If the pipeline is video analytics or canvas, RGB might be a better choice.

**Comment by henbos** [comment]
I'm not trying to create a pixel format picking API, I'm only trying to avoid software MJPEG decoding.

**Comment by henbos** [comment]
> Exposing pixel format selection instead seems to fix the issue and would be useful outside of this edge case.

I believe "SW MJPEG or not?" is the main case, not the edge case. That's when there's a big performance hit.

In theory, you could have different formats, including RGB. But in practise, besides MJPEG, NV12 and YUY2 are the formats exposed by popular webcams and as far as I know NV12 is always better, so the browser can decide.

I am operating under the assumption that pixel format selection is not a significant concern today (@handellm please sanity check, maybe RGB cameras are more popular than I think), but if that is incorrect, my arguments may fall.

> I would not use the term powerEfficient for that API.

Okay, can you come up with a better name? Or are you saying I should abandon this PR?

**Comment by youennf** [comment]
I can see a few choices:
1. A pixelCapturingFormat constraint with values like nv12, I420, mjpeg...
2. A useMJPEG or disallowMJPEG boolean constraint

Both are fine to me.
I tend to prefer the first one as it might be useful to applications using media capture transform.
If we ship 2 and later on want to do 1, we might have to expose two constraints for the same parameter.

Note that in both cases, web applications will need to check the result of getUserMedia and might sometimes have to call applyConstraints to actually opt out of MJPEG.
That is where exposing presets will make web applications life so much easier.

**Comment by henbos** [comment]
I’d prefer to disallow MJPEG (what not to do) than to constrain on a specific format (what to do) since I’m happy to let the browser decide and I don’t think constraints can easily express ”anything but”, so I like disallowMJPEG.

But one thing: if HW decoding is available there’s no reason to avoid it.
- How about disallowSoftwareMJPEG?

**Comment by henbos** [comment]
Alternatively if it was possible to check if HW decoder availability with a different API, one could constrain disallowMJPEG with !hasHW. But disallowMJPEG would be good enough in the meantime

**Comment by youennf** [comment]
> HW decoding is available there’s no reason to avoid it.

HW decoding is still decoding, and is tied to using video frames, not the camera.
Media Capabilities might anyway provide you this information so that the web application can decide what to do.

**Comment by jan-ivar** [comment]
"disallow" implies exact. How about `avoidSoftwareMJPEG`?

**Comment by jan-ivar** [comment]
Also, as much as I like to pick on softwar MJPEG as the next person, I confess I don't follow the arguments for not just calling it `powerEfficient`, as a more abstract expression of the app's concern here.

> For instance {height: 1080, powerEfficient: true} could lead to choosing a frame rate of 1 frame per second?

Yes it could, because in theory `{height: 1080}` could also lead to choosing a frame rate of 1. But is it a realistic concern?

**Comment by henbos** [comment]
> I confess I don't follow the arguments for not just calling it powerEfficient.

I think the argument is to separate ”how to open the camera” from ”how to use the frames” which would in theory be a later step in the pipeline.

In Chrome, decoding happens either inside the capture process or even earlier by the OS, but in theory the implementation could perform decoding much later depending on what API is consuming the frames.

**Comment by henbos** [comment]
In macOS 12 for instance, MJPEG is performed by the OS before it even reaches Chrome where it arrives as NV12. This causes performance issues that we’d like to avoid using this constraint. In this case decoding step and capture step is not entirely separate

**Comment by jan-ivar** [comment]
> In this case decoding step and capture step is not entirely separate

Ok, so wouldn't one option be to scope this efficiency camera constraint to those cases?

> HW decoding is still decoding, and is tied to using video frames, not the camera.
> Media Capabilities might anyway provide you this information so that the web application can decide what to do.

@youennf Ok, so instead of
```js
  const stream = await navigator.mediaDevices.getUserMedia({video: {height: 1080, powerEfficient: true}});
```
...you're proposing apps write the following?
```js
  const isHW = (await navigator.mediaCapabilities.decodingInfo({
    type: 'webrtc',
    video: {
      contentType: 'video/x-motion-jpeg', // or 'video/x-jpeg'?
      width: 1920,
      height: 1080,
      bitrate: 6000000, //?
      framerate: 30
    }
  }).powerEfficient;
  const stream = await navigator.mediaDevices.getUserMedia({video: {
    height: 1080, avoidMJPEG: !isHW
  }});
```
From my reading of the [spec](https://w3c.github.io/media-capabilities/#example1), that's querying _"WebRTC receive capabilities"_ (vs. _"WebRTC send capabilities"_ for `encodingInfo`) which sounds like an RTCPeerConnection source (a `transceiver.receiver.track`), not a local camera.

**Comment by henbos** [comment]
A constraint that tells getUserMedia() to avoid MJPEG decoding would only sensibly be used by an app that does not have an MJPEG-only pipeline. I think this makes it OK not to entirely separate capture from decoding (argument in favor of powerEfficient). But if we can’t do that, avoidMJPEG still solves many issues.

But wrt Jan-Ivar’s comment: I don’t think there currently exists an API for isHW that maps 1:1 to this, because the MJPEG decoder involved here is not from the same set of decoders as MediaCapabilities queries (e.g. if the OS does it for us prior to frames entering the browser this could be separate than what decoders the browser has implemented for other APIs in a pipeline).

**Comment by henbos** [comment]
Show of hands?

Proposal A:
- avoidMJPEG: "Does not open the camera in MJPEG."
- The isHW issue may need to be solved separately to complement it.

Proposal B:
- avoidMJPEG: "Does not open the camera in MJPEG, but only if the browser thinks that decoding MJPEG would negatively impact performance."
- Other name suggestions are welcome... avoidInefficientMJPEG?

Proposal C
- Do both: avoidMJPEG and avoidInefficientMJPEG

**Comment by alvestrand** [comment]
Proposal B gets my vote. I don't care much about names, as long as the description is clear, but avoidInefficientMjpeg has the advantage of being explicit about what it's doing.

**Comment by youennf** [comment]
Oh, so I misunderstood the situation: the MJPEG decoding happens very close to camera level, not downstream and MJPEG frames will never be flowing to the video pipeline.

Also WebCodecs does not have a MJPEG frame, so as soon as we would plug a track transform, the UA would have to convert MJPEG frames to whatever is most appropriate.

I would be tempted to name the constraint to avoidPixelFormatConversion or avoidCompressedPixelFormat, the problem is related to any compressed frame format, not to MJPEG.

**Comment by henbos** [comment]
> Oh, so I misunderstood the situation: the MJPEG decoding happens very close to camera level, not downstream and MJPEG frames will never be flowing to the video pipeline.

That's correct. MJPEG gets decoded regardless.

> I would be tempted to name the constraint to avoidPixelFormatConversion or avoidCompressedPixelFormat, the problem is related to any compressed frame format, not to MJPEG.

Conversions may occur somewhere either at capture or later in the pipeline (NV12 to I420 or RGB or YUY2 to NV12), so `avoidPixelFormatConversion` doesn't quite describe this. But `avoidCompressedPixelFormat` works, since that's what this is about.

@youennf Regarding whether or not "isHW" should be baked into this or a separate API... should we go with `avoidCompressedPixelFormat` or `avoidInefficientCompressedPixelFormat` for now?

**Comment by handellm** [comment]
Going back to the original problem (increased power consumption on MJPEG), I think some variant of avoidInefficientCompressedPixelFormat, powerEfficient, "isHW" makes sense. This is only a problem when there is no hardware support for MJPEG decode.

Saying that, avoidCompressedPixelFormat is also useful as a constraint for quality reasons, (if compressed means lossy in this context).

**Comment by handellm** [comment]
Elaborating a little more on the availability of hardware MJPEG decode - in Chrome this is deployed and used on Windows MediaFoundation and on ChromeOS devices.

**Comment by youennf** [comment]
> This is only a problem when there is no hardware support for MJPEG decode.

Do we have data supporting this?
I would guess that an additional decoding step is always more expensive, not sure by how much, but still.

> in Chrome this is deployed and used on Windows MediaFoundation and on ChromeOS devices.

Does this mean that Chrome would prefer selecting MJPEG pixel format in those platforms to match web app resolution constraints but not in other platforms lacking MJPEG HW decoders?

If so, `avoidInefficientCompressedPixelFormat ` makes some sense.

**Comment by henbos** [comment]
Yes avoiding MJPEG only when there is no HW support is the use case we have in mind.

I updated the pull request to `avoidInefficientCompressedPixelFormat`. Please take a look

File: index.html [hunk 3]
```diff
@@ -208,6 +209,30 @@ <h3>GetUserMediaSemantics enum</h3>
         </table>
       </div>
     </section>
+    <section id="mediatrackconstraintset-dictionary-extensions">
+      <h3>MediaTrackConstraintSet dictionary extensions</h3>
+      <div>
+        <pre class="idl"
+>partial dictionary MediaTrackConstraintSet {
```
[the remark this answers is not in the document]
Comment (henbos) [reply index.html]: Done: I added the constraint to MediaTrackSupportedConstraints, MediaTrackCapabilities and MediaTrackSettings (in addition to MediaTrackConstraintSets).

> getUserMedia should reject a required avoidInefficientCompressedPixelFormat constraint.

Is this the same as adding it to "allowed required constraints for device selection"? I updated the PR to do this, but see other comment.

> It might also be good to write WPT tests as well.

This feature is testable with mocks controlling device capabilities or with real web cams, but AFAIK there's no way to mock device capabilities inside of WPTs? So I think it's currently not possible to test the device selection part unless WPTs are more powerful now than they used to be.

It's possible to WPT test the presence of the constraint though, which would naturally be added when implementing this.

File: index.html [hunk 4]
```diff
@@ -208,6 +211,106 @@ <h3>GetUserMediaSemantics enum</h3>
...
+              the user agent, but decoding MJPEG in software is an example of an
+              expensive mode.</p>
+            </td>
+          </tr>
+        </tbody>
+      </table>
+      <h3>Additions to allowed required constraints for device selection</h3>
+      <p>The following constrainable properties are
+      <a data-cite="mediacapture-streams">allowed required constraints for device selection</a>:
+      <a>avoidInefficientCompressedPixelFormat</a>.
```
Comment (henbos) [inline index.html:311]: 1. Do we really want this in "allowed required constraints for device selection" or should I remove this part?
2. I couldn't get the data citing for this reference to work, the best I got was linking to the correct document. I think the problem is that mediacapture-main's definition of this does not have the "data-export" attribute in the &lt;dfn&gt;, if so we need to do a PR to mediacapture-main as well.

**Review by henbos — COMMENTED** [review]
PTAL

File: index.html (hunk 4 above)
Comment (youennf) [reply index.html:311]: We do not want this in the "allowed required constraints for device selection" list.

Comment (henbos) [reply index.html]: Got it. Removed.

**Comment by henbos** [comment]
Ping. What is the next step? Review and merge or is there still no consensus here?

File: index.html [hunk 5]
```diff
@@ -208,6 +210,99 @@ <h3>GetUserMediaSemantics enum</h3>
...
+        </thead>
+        <tbody>
+          <tr>
+            <td><dfn id="def-constraint-avoidInefficientCompressedPixelFormat">
+            avoidInefficientCompressedPixelFormat</dfn></td>
+            <td>{{ConstrainBoolean}}</td>
+            <td>
+              <p>Compressed pixel formats need to be decoded. The user agent
+              SHOULD label compressed pixel formats that incur significant power
+              penalty when decoded as not power efficient. The labeling is up to
```
Comment (youennf) [inline index.html:298]:
```suggestion
              penalty when decoded as inefficient. The labeling is up to
```

File: index.html [hunk 6]
```diff
@@ -208,6 +210,99 @@ <h3>GetUserMediaSemantics enum</h3>
...
+            <th>Notes</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td><dfn id="def-constraint-avoidInefficientCompressedPixelFormat">
+            avoidInefficientCompressedPixelFormat</dfn></td>
+            <td>{{ConstrainBoolean}}</td>
+            <td>
+              <p>Compressed pixel formats need to be decoded. The user agent
```
Comment (youennf) [inline index.html:296]:
```suggestion
              <p>Compressed pixel formats often need to be decoded, for instance for display purposes or when being encoded during a video call. The user agent
```

File: index.html [hunk 7]
```diff
@@ -208,6 +210,99 @@ <h3>GetUserMediaSemantics enum</h3>
...
+            <th>Property Name</th>
+            <th>Values</th>
+            <th>Notes</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td><dfn id="def-constraint-avoidInefficientCompressedPixelFormat">
+            avoidInefficientCompressedPixelFormat</dfn></td>
+            <td>{{ConstrainBoolean}}</td>
```
Comment (youennf) [inline index.html:294]: Do we agree that a lack of an avoidInefficientCompressedPixelFormat constraint is equivalent to avoidInefficientCompressedPixelFormat=false?
Should we state it explicitly?

**Review by youennf — APPROVED** [review]

File: index.html [hunk 8]
```diff
@@ -208,6 +210,104 @@ <h3>GetUserMediaSemantics enum</h3>
...
+          <tr>
+            <th>Property Name</th>
+            <th>Values</th>
+            <th>Notes</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td><dfn id="def-constraint-avoidInefficientCompressedPixelFormat">
+            avoidInefficientCompressedPixelFormat</dfn></td>
```
Comment (jan-ivar) [inline index.html:293]: This will show up in `track.getSettings()`, which I like. But as a name, having a verb associated with this property doesn't make a lot of sense to me.

[Constrainable properties](https://w3c.github.io/mediacapture-main/#defining-a-new-constrainable-property) are usually not verbs (yes I'm aware I introduced "avoid"), but properties of the source that show up in constraints, capabilities _and_ settings. This makes me think that a better name would be `efficientCompressedPixelFormat` or `powerEfficient`. E.g.

```js
if (!track.getSettings().powerEfficient) {
  await track.applyConstraints({powerEfficient: true});
}
```
...or
```js
if (!track.getSettings().efficientCompressedPixelFormat) {
  await track.applyConstraints({efficientCompressedPixelFormat: true});
}
```

**Review by jan-ivar — CHANGES_REQUESTED** [review]

File: index.html (hunk 8 above)
Comment (jan-ivar) [reply index.html:293]: Wait, that's not the correct opposite of `avoidInefficientCompressedPixelFormat`... `efficientOrUncompressedPixelFormat`? `efficientPixelFormat`? ugh.

Can we just go with `powerEfficient`?

Comment (henbos) [reply index.html:293]: Fine by me. @youennf does your LGTM still hold if we call it powerEfficient?

Comment (youennf) [reply index.html:293]: powerEfficient is a very broad term for camera. We might in the future want to add yet another powerEfficiency related property.
Given we nailed down the particular thing we want to constraint, I prefer more focused terms.

Also, we already have verb properties in getDisplayMedia so verbs are fine if they are more understandable.
I would stick with avoidInefficientCompressedPixelFormat, defaulting to false.
inefficientCompressedPixelFormat='include|exclude' might be okayish in the light of recent gDM preferences.

Comment (henbos) [reply index.html:293]: I think "avoid" works both as a constraint (I would or would not like to avoid...) and as a setting (the current setting does or does not avoid). But this does feel very much like bike shedding. @jan-ivar

Comment (youennf) [reply index.html:293]: powerEfficientPixelFormat is okayish too. The downsides is that a powerEfficientPixelFormat=false constraint seems to mean you want to select a non efficient pixel format while we want it to mean any pixel format is ok (distance is always the same). Maybe this is a bad idea though, when exposed as a setting you probably prefer to know whether the format is actually power efficient or not.

Comment (henbos) [reply index.html:293]: powerEfficientPixelFormat which is true or false depending on whether or not it is efficient makes a lot of sense to me, I'd just have to remove the sentence that says that false is the same as the default. I agree the setting should reflect if the current setting is or is not efficient.

Comment (jan-ivar) [reply index.html:293]: Selecting a non-efficient pixel format seems useful for testing. It'd allow measuring the cost-benefit of forgoing these modes (which presumably have other decent qualities or they wouldn't exist).

Comment (henbos) [reply index.html]: Okay I changed it to powerEfficientPixelFormat. Is this good enough to merge?

Comment (youennf) [reply index.html]: Let's validate this during today's editor call.

Comment (henbos) [reply index.html]: Is there a conclusion?

Comment (youennf) [reply index.html]: It was brought that we should probably discuss this during a WebRTC interim.
This was discussed at yesterday's WebRTC interim meeting and there is consensus to move forward.
I am guessing we will be able to finalise this at next editor's call.

Comment (henbos) [reply index.html]: OK, thank you :)

File: index.html [hunk 9]
```diff
@@ -208,6 +210,107 @@ <h3>GetUserMediaSemantics enum</h3>
         </table>
       </div>
     </section>
+    <section id="mediatracksupportedconstraints-dictionary-extensions">
```
Comment (alvestrand) [inline index.html:213]: I think there should be a </section> <section> and a <h2> here to separate the addition of "powerEfficient" from the new picker model section.

**Comment by youennf** [comment]
Other PRs are adding new constraints in their own section which seems more tractable.
